### PR TITLE
Multilevel crates

### PR DESCRIPTION
**Pull request by vlada-dudr**
Hi all, please review the code, write suggestions or ideas. There are still some bugs I know about, which I will fix and features I will add. I'd love to have some feedback before I finish the whole thing.

Next steps:
- deleting folders recursively
- fix - closing folders isn't removing them from list of opened folders
- moving crates between folders
- filling folder screen somehow

### CODE REVIEW

File: src/library/cratefeature.h [hunk 1]
```diff
@@ -49,8 +51,12 @@ class CrateFeature : public LibraryFeature {
     void activateChild(const QModelIndex& index);
     void onRightClick(const QPoint& globalPos);
     void onRightClickChild(const QPoint& globalPos, QModelIndex index);
+    void onLazyChildExpandation(const QModelIndex& index);
+    void itemCollapsed(const QModelIndex& index);
 
-    void slotCreateCrate();
+    void slotCreateCrate(int crateType);
+    inline void slotCreateNormalCrate() { slotCreateCrate(0); } ;
```
Comment (daschuer) [inline src/library/cratefeature.h]: What is Normal? -> slotCreateTrackCrate

File: src/library/cratefeature.cpp [hunk 2]
```diff
@@ -110,6 +124,7 @@ CrateFeature::~CrateFeature() {
 #ifdef __AUTODJCRATES__
     delete m_pAutoDjTrackSource;
 #endif // __AUTODJCRATES__
+    delete m_pCreateCrateFolderAction;
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: This and the delete above are redundant because it will be deleted bay the Qt Object tree, since it is parented to this.

File: src/library/cratefeature.cpp [hunk 3]
```diff
@@ -128,17 +143,71 @@ int CrateFeature::crateIdFromIndex(QModelIndex index) {
...
+    while (! cratesToProcess.empty()) {
+        crate = cratesToProcess.pop();
+
+        if (id == crateIdFromIndex(crate)) {
+            qDebug() << "Found index of crate found id: " << id;
+            return crate;
+        }
+
+        for(int row = 0; row < m_childModel.rowCount(crate); row++) {
+            cratesToProcess.push(m_childModel.index(row, 0, crate));
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: This can be improved by using a recursive function. This helps to get rid of the "cratesToProcess" stack.

File: src/library/cratefeature.cpp [hunk 4]
```diff
@@ -128,17 +143,71 @@ int CrateFeature::crateIdFromIndex(QModelIndex index) {
...
+}
+
+// updates hasChildren function in childModel
+void CrateFeature::refreshModelHasChildren() { 
+    //put the set of folders to variable to be lambda-captureable
+    QSet<int> folderSet = m_crateDao.getFolderSet();
+    m_childModel.setHasChildrenFunction( 
+        [folderSet] (QVariant crateId) {
+             return folderSet.contains(crateId.toInt());
+        }
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: Year, lamdas a pretty cool, but It looks like we do not need them in this context.
Cant we just pass the folderSet and put the Lamda content directly to m_childModel? This would streamline the code.

File: src/library/cratefeature.cpp [hunk 5]
```diff
@@ -252,15 +322,81 @@ void CrateFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index)
     menu.exec(globalPos);
 }
 
-void CrateFeature::slotCreateCrate() {
+
+void CrateFeature::onLazyChildExpandation(const QModelIndex& index) {
+    qDebug() << "\e[31mEntering CrateFeature::onLazyChildExpandation\e[0m";
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: Is this a color tag? Be sure to remove it before merge.

File: src/library/cratefeature.cpp [hunk 6]
```diff
@@ -128,17 +143,71 @@ int CrateFeature::crateIdFromIndex(QModelIndex index) {
...
+        }
+    );
+}
+
+void CrateFeature::slotCrateTableChanged(int crateId) {
+    qDebug() << "slotCrateTableChanged() playlistId:" << crateId;
+
+    onLazyChildExpandation(QModelIndex());
+    QModelIndex index;
+    foreach(int folderId, m_openFolders) {
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: Please add a comment what this loop does and why

File: src/library/cratefeature.cpp [hunk 7]
```diff
@@ -222,9 +293,7 @@ void CrateFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index)
     }
 
     bool locked = m_crateDao.isCrateLocked(crateId);
-
-    m_pDeleteCrateAction->setEnabled(!locked);
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: Why was this removed?

File: src/library/cratefeature.cpp [hunk 8]
```diff
@@ -299,7 +446,7 @@ void CrateFeature::slotDeleteCrate() {
     if (crateId == -1) {
         return;
     }
-
+    //TODO(vlada-dudr) should we check lock of parents?
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: Yes

File: src/library/cratefeature.cpp [hunk 9]
```diff
@@ -322,6 +469,7 @@ void CrateFeature::slotRenameCrate() {
     }
     QString oldName = m_crateDao.crateName(crateId);
 
+    //TODO(vlada-dudr) should we check lock of parents?
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: Yes

File: src/library/cratefeature.cpp [hunk 10]
```diff
@@ -722,16 +769,19 @@ void CrateFeature::slotTrackSelected(TrackPointer pTrack) {
...
+    cratesToProcess.push(rootItem);
+    while (! cratesToProcess.empty()) {
+        crate = cratesToProcess.pop();
+        
+        crate->setBold(
+                m_cratesSelectedTrackIsIn.contains(
+                    crate->dataPath().toInt()));
+
+        for(int row = 0; row < crate->childCount(); row++) {
+            cratesToProcess.push(crate->child(row));
```
Comment (daschuer) [inline src/library/cratefeature.cpp]: This might be likely a performance killer on low end hardware. 
Use a recursive lookup without a temporary stack

**Comment by daschuer** [comment]
Thank you for all the work. 
I have just skimmed though the code and added some comments. 
It looks promising.

File: src/library/cratefeature.cpp (hunk 4 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: It is not straigh way how to do it. I saw generally two other ways of doing the trick:
1) construct full childmodel at once, no lazy loading -> no need to implement hasChildren function
2) derive treeItemModel to accept list of folders and hard code hasChildren function inside

The solution I used allows any outer object to set how the presence of children is resolved without need to derive treeitemmodel for different usage.

It you think it is better to use other options I will do it that way

File: src/library/cratefeature.cpp (hunk 5 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: I'm planning to remove most of the qDebug lines when it is resolved working

File: src/library/cratefeature.cpp (hunk 4 above)
Comment (daschuer) [inline src/library/cratefeature.cpp]: Ah, now I understand. The lambda makes sense here, since we are in the common tree.

File: src/library/cratefeature.cpp (hunk 2 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: So I am removing all the QAction deletes from destructor.

Comment (daschuer) [inline src/library/cratefeature.cpp]: Yes, thank you.

File: src/library/cratefeature.cpp (hunk 3 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: Ok

File: src/library/cratefeature.h (hunk 1 above)
Comment (vlada-dudr) [inline src/library/cratefeature.h]: NormalCrate is crate. Not folder not anything else.
This function is just shortcut. It is possible to everywhere instead of slotCreateNormalCrate() slotCreateCrate(0)

Comment (daschuer) [inline src/library/cratefeature.h]: I would prefer "TrackCrate"

File: src/library/cratefeature.cpp (hunk 6 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: Yes

File: src/library/cratefeature.cpp (hunk 7 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: I'll add more complex handling of this. We probably need to check if there is any children locked.

File: src/library/cratefeature.cpp (hunk 10 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: done

File: src/library/cratefeature.cpp (hunk 6 above)
Comment (vlada-dudr) [inline src/library/cratefeature.cpp]: done

**Comment by daschuer** [comment]
I have just tested this PR and it is great to see the crates tree. However there are a lot of tiny things to do, you probably know. So what is the mode to continue here? Is there anything special to test? 

One jumping point is surely the issue we have discussed in https://bugs.launchpad.net/mixxx/+bug/671632: Can a crate contain crates and tracks?

**Comment by uklotzde** [comment]
Unfortunately this work has become somehow redundant compared to the more ambitious approach implemented in [PR 1304 Nested Crates](https://github.com/mixxxdj/mixxx/pull/1304).
